### ISSUE TITLE
Porting functionality from LAMMPS for inorganic materials

### ISSUE THREAD

**Issue by philipturner** [openmm/openmm#3929]
Spinoff discussion from https://github.com/openmm/openmm/issues/3924

This issue is to discuss whether we can port the AIREBO force field from LAMMPS (https://arxiv.org/pdf/1810.07026.pdf) and the feasibility of simulating million-atom dense covalent structures. OpenMM's current built-in forces are specialized for biomolecules, but we can work to enable material-science use cases.

I want to use this functionality to test machinery theorized in <i>Nanosystems (Drexler, 1992)</i>. He needed stiff diamond machines to manipulate extremely reactive molecules with ~0.01-0.1 nm precision. There is much controversy over whether designs are just speculation, or viable in reality. Back in 1992, computers struggled to simulate 100 atoms with molecular dynamics. Now we can readily run 100,000-atom simulations with the same speed. OpenMM's cross-platform GPU acceleration pushes the boundaries of testable system size, while being easier to extend than alternative code bases.

Despite this processing power, we still need creative techniques to decrease simulation time to something viable. One idea - redesign supermassive diamondoid structures to have mostly hollow interiors. Using the R^4 law for pipe strength, this achieves the rigidity of sapphire while decreasing thickness tenfold (5x less atoms). In addition, I need to understand how coarse-grained simulations work, and how to use them without sacrificing accuracy at critical contact points.

**Comment by peastman** [comment]
> This issue is to discuss whether we can port the AIREBO force field from LAMMPS

Take a look at [CustomManyParticleForce](docs.openmm.org/latest/api-python/generated/openmm.openmm.CustomManyParticleForce.html).  Can it do what you need?

> I want to use this functionality to test machinery theorized in Nanosystems (Drexler, 1992).

Can you provide a full citation?

You're trying to do a few things that are perhaps not compatible with each other.  You want to simulate a very large system, but you also want to use an advanced many particle potential function.  Many particle potentials tend to be very expensive.  Consider an atom with 10 other atoms within the cutoff distance.  There are 10 possible 2-body terms, 100 possible 3-body terms, 1000 possible 4-body terms...  You see the problem.  In practice, many particle potentials need to be used with a very short cutoff, but even then they're more expensive than the simpler pairwise ones.  Given the size of the system you want to simulate, you probably are better off sticking to something fast.

I don't have experience with materials science simulations of the sort you describe.  The basic simulation machinery should all be the same, but the potential functions will be different.  The place to start will be identifying suitable potentials you could use.

**Comment by philipturner** [comment]
> an advanced many particle potential function

I'm honestly not even sure we need this function. I think the LAMMPS guy was just trying to scare me away from GPU acceleration. I rewatched [Drexler's talk](https://www.youtube.com/watch?v=HjgjtAk-lws&list=PLH78wfbGI1x3zNH-uA6RHNsf8AZnxTErI&index=30), the important part is 17:03 - 20:03. Drexler mentioned LAMMPS but I couldn't find AIREBO or Tersoff in the transcript.

> Can you provide a full citation?

The 1992 book is not free (I borrowed it from a library), but he published a [slightly dated draft](https://dspace.mit.edu/handle/1721.1/27999) in 1991 as his MIT thesis. The old version doesn't include the material in question, but the proper citation is (<i>Nanosystems, 1992, Figure 16.4</i>). Robert Freitas gives a good description in [<i>Nanomedicine Vol. A, 9.3.1.5, Figure 9.10</i>](http://www.nanomedicine.com/NMI/9.3.1.5.htm). Perhaps we refrain from attempting such a large machine, for now.

> The place to start will be identifying suitable potentials you could use.

Drexler details the molecular dynamics algorithms he used c. 1991. Section 3.3 of the thesis (pages 55-68) talks about an "MM2" software package. I couldn't comprehend it on the first read through, but he seems to list all the formulae. He simulated a ~100 atom bearing (thesis Figure 1.1, page 15) using MM2. Is it possible to start by reproducing that?

If OpenMM does correctly simulate diamond, I could try the "lock" depicted in ([IMM Report 46, Page 22, Figure 15 & 16](http://www.imm.org/Reports/rep046.pdf)). I eyeballed it at 15,000 atoms and we need 21 of these (~300k atoms) for a reversible logic gate. I want to investigate the resonance at 18 GHz (0.06 ns), and the 100 MHz clock frequency (10 ns).

**Comment by philipturner** [comment]
I found the [original research paper](https://aip.scitation.org/doi/pdf/10.1063/1.481208) for AIREBO. Established in 2000, it improves upon an earlier REBO (reactive empirical bond-order) model. REBO allows for breaking and re-forming covalent bonds, therefore making the bond order dynamic. The most expensive term is probably torsion, a four-atom bonded interaction. It's used to study chemical vapor deposition of diamond, carbon nanotubes, and tribology of diamond interfaces.

It would be useful for studying the interface between moving diamondoid parts. Also, atom-by-atom mechanosynthesis is similar to vapor deposition. This seems to be a much more efficient alternative to density functional theory.

However, it's a potential suited for highly reactive environments. Bulk diamond it mostly static, especially the interior of large parts. For example, most surface area on the "lock" from IMM has no potential for chemical reactions. Only maybe the breakdown of acetylene joints\*, <1% of all atoms. For the bulk component, I'd care mostly about propagation of sound waves that incite destructive resonance.

I discussed a lot of subjects here, so to clarify:
1) Figure out what Drexler did in 1990.
2) Successfully reproduce his simulations of ~100-atom bearings.
3) Port the REBO potential using CustomManyParticleForce, cutoff = carbon covalent bond radius.
4) Port the non-bonded and inter-molecular forces from the upgraded AIREBO model.
5) Find an accurate way to apply AIREBO inside predefined volumetric zones, and simpler potentials elsewhere.
6) Model one "lock" from the IMM report (~15,000 atoms).
7) Model multiple mechanical locks\* (~120-315,000 atoms).
8) Model medium-sized systems centered around a tiny zone of massive reactivity (the mechanosynthesis site).
9) See whether simulating a 80% hollow Stewart platform (~600,000 atoms) is possible.

<details><summary>* Image of a joint (high-reactivity zone)</summary>

<img width="478" alt="Screenshot 2023-01-24 at 10 29 15 AM" src="https://user-images.githubusercontent.com/71743241/214336277-88d69c12-8037-49e9-b1db-d2eea6ebdb7f.png">

</details>

**Comment by peastman** [comment]
Here's the reference for the MM2 force field: https://pubs.acs.org/doi/10.1021/ja00467a001.  Published in 1977.  That was a really early one!

You may need to use different models to study different things.  If you want to study sound wave propagation, you don't want to make it hollow.  That would totally change the behavior.  But you don't need a reactive potential to study that.  For surface phenomena you're more likely to care about reactivity, but for that it's probably fine to make the object rigid and hollow out the interior.  But don't take any of my advice on this too seriously.  You need to talk to someone who has experience simulating this sort of system.

**Comment by philipturner** [comment]
I now understand the requirements. They are very strange and nothing like biomolecules.

Machine-phase chemistry is like a different phase of matter - very, very stiff. It's designed to behave predictably, be unreactive, and resist energy fluctuations. I quantatively compared the energy deviations from using FP32 vs. FP64, and they pale in comparison to the <b>energy gradients</b> (e.g. strain forces) that nanomachines face. In biomolecules, energy is crucial to determine reaction rates. That is not true for Drexler's machines.

Second, I can get away with very simple forcefields. Potentially 4 fs timesteps too. Tersoff is used for highly reactive situations, typically found in solution-phase chemistry. Nanomachines are designed to not suffer from uncontrollable reactions. There are highly reactive moieties - carbon radicals - in mechanosynthesis, but those are more appropriately simulated through first-principles DFT. The radicals are also highly controlled, held in place by quite impressive forces.

In short, GPUs are unexpectedly well-suited for this new kind of chemistry, and I don't even have to port complex forcefields to simulate it.

**Comment by maaku** [comment]
MM2 is an outdated force field. But so is the Universal Force Field supported by LAMMPS. Here is a more recent universal force field that supports inorganic ligands and gives much better (more realistic) results. It is what I would recommend using for this application.

https://onlinelibrary.wiley.com/doi/full/10.1002/anie.202004239

**Comment by philipturner** [comment]
I want to reproduce Eric Drexler's work from Nanosystems. In programming, you often narrow a bug by exactly reproducing the conditions that caused it. Change one variable, and it could disappear. This is how I feel with Nanosystems - I <i>will</i> come across bugs and problems reproducing things. My thinking is, by using the MM2 forcefield, I can eliminate one extra confounding variable when I try to solve those bugs.

GFN-FF[^1] seems very promising for structures that lack much empirical data. It is $O(n^2)$ just like nonbonded forces, but I wonder whether it can use a cutoff. Cutoffs are necessary to scale beyond ~10K atoms and toward ~1M atoms. I also need to know how expensive the $O(n^2)$ part is compared to LJ and Coulomb interactions.

SHAKE is necessary to use larger timesteps like 4 fs (instead of 1 fs) and simulate nanomachines in real-time[^2]. It also emulates quantum suppression of high-frequency motion, making it more realistic[^3]. GFN-FF [implements SHAKE](https://github.com/grimme-lab/xtb/blob/main/src/gfnff/shake_module.f90), even though QSHAKE[^4] is more accurate for rigid systems.

[^1]: Source code: https://github.com/grimme-lab/xtb/tree/main/src/gfnff

[^2]: I achieved 60,000 ns/day for several hundred atoms using only LJ; that corresponds to 700 ps/s. The more expensive the nonbonded force, the fewer the atoms can be simulated at this speed. Ref: https://github.com/philipturner/noble-gas-md-simulation

[^3]: Appendix A.6, "Evaluating the friction of rotary joints in molecular machines": https://pubs.rsc.org/en/content/articlelanding/2017/me/c7me00021a

[^4]: Quaternion SHAKE algorithm, a relatively simple extension of SHAKE: https://onlinelibrary.wiley.com/doi/10.1002/(SICI)1096-987X(19980115)19:1%3C102::AID-JCC9%3E3.0.CO;2-T

**Comment by maaku** [comment]
You are not going to need to scale beyond ~10k atoms, certainly not 1M atoms. You are talking about diamondoid mechanical systems operating in vacuum, right? Extremely rigid structures fixed in place and undergoing constrained motions are not going to interact appreciably with any atoms other than the those of the parts it comes in direct contact with. To check if a gear has the properties you want to to have (e.g. near frictionless motion) you need to simulate the gear structure, the shaft it is attached to, and the other interlocking gear, and that's it. The van der Waals forces diamondoid structures doesn't have any appreciable affect beyond a few atomic radii. Once you know how the part works you can proceed to model it as an aggregate part with more traditional Newtonian properties.

**Comment by philipturner** [comment]
I would like to test a nanomechanical computer, and test the validity of coarse-graining algorithms. Atomistic detail enables insights such as studying sound wave propagation, examining resonance modes, and greater accuracy. I can still reach very long simulation times - several 1 GHz clock cycles - by running an overnight simulation. While it's not the only way to simulate such large systems, it could be a useful tool.

What are some methods for coarse-graining these structures, and what speedups do they promise?

**Comment by maaku** [comment]
I'm not sure I understand--you would model it the same way an engineer models macro-scale parts. Calculate its moment of inertia, resonant modes, frictional heating, Young's modulus, etc. etc.  You'd need atomistic detail of both the part and its interconnects in order to simulate these properties. But once you've done that you treat the part as if it was its actual macro-scale equivalent. Treat a molecular gear the same way you would a cut metal gear in the gearbox of a car (except that lubricant wouldn't be required), or the gemstone gear in a Rolex watch. You calculate moments of inertia, gear ratios, response times, critical damping coefficients, etc. Look in the textbook for any junior-year Engineering Dynamics course.

(I'd be happy to take this discussion elsewhere if necessary. We've wandered off the topic of this issue, and out of scope for this project.)

**Comment by maaku** [comment]
To expand a little on my previous comment a mechanical engineer designing a clockwork mechanism, like a watch or a Babbage computer, doesn't actually do any time-domain simulations, coarse-grained or otherwise. Rather they do calculations such as: "given the gear ratios involved, how many revolutions of the main axel are required to complete each logical gate operation?", "what's the longest sequence of gates in series?", "is there sufficient time between ticks to complete all necessary calculations, such that the interface rods end up back at the ready state?" etc. Also, "given each start or stop motion generates X eV of thermal energy when running at N Ghz, how much thermal energy is produced? is that heat flux within the known conductive limits of the surrounding machinery, such that the heat will have time to dissipate?" etc.

These are calculations that must be performed to validate a design, but they are not simulations. The simulations only need to be done part-wise to determine what the inertial, thermal, resonant, frictional, etc. properties of the part are, and then you use those properties to validate a design using the traditional tools of mechanical engineering.
